File: verify/iscostas.hs
import Data.List (nub)

-- TODO: documentation

iscostas :: [Int] -> Bool
iscostas perm =
    length (nub perm) == n &&
    all checkShift [1 .. n - 1]
  where
    n = length perm
    checkShift k =
        let diffs = zipWith (-) (drop k perm) perm
        in length (nub diffs) == length diffs

main :: IO ()
main = do
    _ <- getLine
    line2 <- getLine
    let perm = map read (words line2) :: [Int]
    putStrLn $ if iscostas perm then "1" else "0"

File: verify/iscostas.f90
! TODO: documentation

program iscostas
    implicit none
    integer :: n, i, k, d
    integer, allocatable :: perm(:), seen(:)
    logical :: costas

    read (*, *) n
    allocate (perm(n), seen(2 * n + 1))
    read (*, *) perm

    costas = .true.

    seen = 0
    do i = 1, n
        if (perm(i) < 1 .or. perm(i) > n .or. seen(perm(i)) == 1) then
            costas = .false.
        end if
        seen(perm(i)) = 1
    end do

    if (costas) then
        do k = 1, n - 1
            seen = 0
            do i = 1, n - k
                d = perm(i + k) - perm(i) + n + 1
                if (seen(d) == 1) then
                    costas = .false.
                    exit
                end if
                seen(d) = 1
            end do
            if (.not. costas) exit
        end do
    end if

    if (costas) then
        write (*, '(i1)') 1
    else
        write (*, '(i1)') 0
    end if

    deallocate (perm, seen)
end program iscostas
